backend/src/auth/auth.guard.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  Injectable,
+  CanActivate,
+  ExecutionContext,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import jwt from 'jsonwebtoken';
+
+import { AUTH_SERVER, AUTH_REALM } from './const';
+import { Token } from './interface';
+import { Metadata } from './metadata';
+
+@Injectable()
+export class AuthGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
+
+  canActivate(context: ExecutionContext): boolean {
+    const isPublic = this.reflector.getAllAndOverride<boolean>(
+      Metadata.IS_PUBLIC,
+      [context.getHandler(), context.getClass()],
+    );
+
+    if (isPublic) {
+      return true;
+    }
+
+    const request = context.switchToHttp().getRequest();
+    const authHeader = request.headers.authorization;
+    const token = this.parseJwt(authHeader);
+
+    if (!authHeader.includes('Bearer ')) {
+      throw new UnauthorizedException('Unauthorized user without credentials');
+    }
+
+    if (!token) {
+      throw new UnauthorizedException();
+    }
+    try {
+      this.validateToken(token);
+    } catch {
+      throw new UnauthorizedException();
+    }
+    return true;
+  }
+
+  parseJwt(token: string): Token {
+    const base64Url = token.split('.')[1];
+    const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+    const jsonPayload = decodeURIComponent(
+      atob(base64)
+        .split('')
+        .map((c) => {
+          return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+        })
+        .join(''),
+    );
+
+    return JSON.parse(jsonPayload);
+  }
+
+  validateToken(token: Token) {
+    if (!token) {
+      throw new UnauthorizedException();
+    }
+    const payload = token as jwt.JwtPayload;
+
+    if (payload.iss !== `${AUTH_SERVER}/realms/${AUTH_REALM}`) {
+      throw new UnauthorizedException();
+    }
+
+    if (payload.aud !== 'account') {
```

Comment:
This seems like the wrong thing to check? I think authorized party is the thing to check `azp`